gno.land/pkg/sdk/vm/builtins.go

```diff
@@ -60,21 +65,82 @@ func (bnk *SDKBanker) RemoveCoin(b32addr crypto.Bech32Address, denom string, amo
 // SDKParams
 
 type SDKParams struct {
-	vmk *VMKeeper
+	pmk *params.ParamsKeeper
 	ctx sdk.Context
 }
 
-func NewSDKParams(vmk *VMKeeper, ctx sdk.Context) *SDKParams {
+// These are the native function implementations bound to standard libraries in Gno.
+// All methods of this struct are not intended to be called from outside vm/stdlibs/std.
+//
+// The key has the format <realm>.<paramname>.<type>:
+// realm: A realm path indicating where Set methods are called from.
+// paramname: The parameter key. If it contains a prefix that matches the module's paramkey
+// prefix (which by default is the module name), it indicates an attempt to set the module's
+// parameters for the chain. Otherwise, it is treated as an arbitrary parameter.
+// type: The primitive type of the parameter value.
+
+func NewSDKParams(pmk *params.ParamsKeeper, ctx sdk.Context) *SDKParams {
 	return &SDKParams{
-		vmk: vmk,
+		pmk: pmk,
 		ctx: ctx,
 	}
 }
 
-func (prm *SDKParams) SetString(key, value string)      { prm.vmk.prmk.SetString(prm.ctx, key, value) }
-func (prm *SDKParams) SetBool(key string, value bool)   { prm.vmk.prmk.SetBool(prm.ctx, key, value) }
-func (prm *SDKParams) SetInt64(key string, value int64) { prm.vmk.prmk.SetInt64(prm.ctx, key, value) }
-func (prm *SDKParams) SetUint64(key string, value uint64) {
-	prm.vmk.prmk.SetUint64(prm.ctx, key, value)
+func (prm *SDKParams) SetString(key gstd.ParamKey, value string) {
+	prm.assertRealmAccess(key)
+	prm.willSetKeeperParams(prm.ctx, key, value)
+	prm.pmk.SetString(prm.ctx, key.String(), value)
+}
+
+// Set a boolean parameter in the format of realmPath.parameter.bool
+func (prm *SDKParams) SetBool(key gstd.ParamKey, value bool) {
+	prm.assertRealmAccess(key)
+	prm.willSetKeeperParams(prm.ctx, key, value)
+	prm.pmk.SetBool(prm.ctx, key.String(), value)
+}
+
+func (prm *SDKParams) SetInt64(key gstd.ParamKey, value int64) {
+	prm.assertRealmAccess(key)
+	prm.willSetKeeperParams(prm.ctx, key, value)
+	prm.pmk.SetInt64(prm.ctx, key.String(), value)
+}
+
+func (prm *SDKParams) SetUint64(key gstd.ParamKey, value uint64) {
+	prm.assertRealmAccess(key)
+	prm.willSetKeeperParams(prm.ctx, key, value)
+	prm.pmk.SetUint64(prm.ctx, key.String(), value)
+}
+
+func (prm *SDKParams) SetBytes(key gstd.ParamKey, value []byte) {
+	prm.assertRealmAccess(key)
+	prm.willSetKeeperParams(prm.ctx, key, value)
+	prm.pmk.SetBytes(prm.ctx, key.String(), value)
+}
+
+// ParamKey's prefix must match the keeper's paramKeyPrefix; otherwise it will panic and revert the transaction.
+func (prm *SDKParams) willSetKeeperParams(ctx sdk.Context, key gstd.ParamKey, value interface{}) {
+	kp := key.Prefix
+	// ParamfulKeeper can be accessed from SysParamsRealmPath only
+	if key.Realm != SysParamsRealmPath || kp == "" {
```

Comment:
redundant with assertRealmAccess.

Reply:
Yes, this check ensures that willSetKeeperParams can be called independently from assertRealmAccess. There should be no sequential dependency between these two methods.

Reply:
They are always called together and perform the same check, though it is written in very different ways. I wouldn't care that much being ineffective in terms of performance since this situation should be relatively rare. However, the redundancy makes the code less maintainable, as it employs two different methods to perform the same check, which are always called together.

Reply:
Can't we have two independent functions that eventually call a shared helper (or not)? I find what we have here quite confusing.